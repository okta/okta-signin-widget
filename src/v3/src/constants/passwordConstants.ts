/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

export const PASSWORD_REQUIREMENTS_KEYS: Record<'complexity' | 'age', { [key: string]: string }> = {
  complexity: {
    minLength: 'password.complexity.length.description',
    minLowerCase: 'password.complexity.lowercase.description',
    minUpperCase: 'password.complexity.uppercase.description',
    minNumber: 'password.complexity.number.description',
    minSymbol: 'password.complexity.symbol.description',
    excludeUsername: 'password.complexity.no_username.description',
    excludeFirstName: 'password.complexity.no_first_name.description',
    excludeLastName: 'password.complexity.no_last_name.description',
    useADComplexityRequirements: 'password.complexity.adRequirements.description',
  },
  age: {
    minAgeMinutes: 'password.complexity.minAgeMinutes.description',
    minAgeHours: 'password.complexity.minAgeHours.description',
    minAgeDays: 'password.complexity.minAgeDays.description',
  },
};
