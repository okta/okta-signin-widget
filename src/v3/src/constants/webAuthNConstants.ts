/*
 * Copyright (c) 2024-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IDX_STEP } from './idxConstants';

export const ABORT_REASON_CLEANUP = 'WebAuthNAutofill component cleanup';
export const ABORT_REASON_WEBAUTHN_AUTOFILLUI_STEP_NOT_FOUND = `${IDX_STEP.CHALLENGE_WEBAUTHN_AUTOFILLUI_AUTHENTICATOR} not found in available steps`;
