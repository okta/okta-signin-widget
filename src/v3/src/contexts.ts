/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { createContext } from 'preact';
import { useContext } from 'preact/hooks';

import {
  ILayoutContext, IStepperContext, IWidgetContext, UISchemaLayoutType,
} from './types';

const createWidgetContext = <T extends unknown>() => {
  // Create a context with a generic parameter or undefined
  const genericContext = createContext<T | undefined>(undefined);

  // Check if the value provided to the context is defined or throw an error
  const useWidgetContext = () => {
    const contextIsDefined = useContext(genericContext);
    if (!contextIsDefined) {
      throw new Error('useWidgetContext must be used within a Provider');
    }
    return contextIsDefined;
  };

  return [useWidgetContext, genericContext.Provider] as const;
};

const [useWidgetContext, WidgetContextProvider] = createWidgetContext<IWidgetContext>();

export { useWidgetContext, WidgetContextProvider };

export const StepperContext = createContext<IStepperContext>({
  stepIndex: 0,
  setStepIndex: () => {},
});

export const useStepperContext = () => useContext(StepperContext);

export const LayoutContext = createContext<ILayoutContext>({
  layoutDirection: UISchemaLayoutType.VERTICAL,
});

export const useLayoutContext = () => useContext(LayoutContext);
