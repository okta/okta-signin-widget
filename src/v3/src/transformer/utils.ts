/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  FieldElement,
  FormBag,
  UISchemaElement,
  UISchemaLayoutType,
} from '../types';

export const createForm = (): FormBag => ({
  schema: {},
  uischema: {
    type: UISchemaLayoutType.VERTICAL,
    elements: [],
  },
  data: {},
  dataSchema: {
    fieldsToValidate: [] as string[],
  } as FormBag['dataSchema'],
});

export const removeUIElementWithName = (
  name: string,
  elements: UISchemaElement[],
): UISchemaElement[] => (
  elements.filter((element) => (
    name !== (element as FieldElement).options?.inputMeta?.name
  ))
);

export const getUIElementWithName = (
  name: string,
  elements: UISchemaElement[],
): UISchemaElement | undefined => (
  elements.find((element) => (
    name === (element as FieldElement).options?.inputMeta?.name
  ))
);
