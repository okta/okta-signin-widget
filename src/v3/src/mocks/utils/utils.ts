/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxStatus, IdxTransaction } from '@okta/okta-auth-js';
import { IDX_STEP } from 'src/constants';
import { DataSchemaBag, FormBag, UISchemaLayoutType } from 'src/types';

export const getStubTransaction = (status: IdxStatus = IdxStatus.PENDING): IdxTransaction => ({
  status,
  proceed: jest.fn(),
  neededToProceed: [],
  rawIdxState: {
    version: '',
    stateHandle: '',
  },
  actions: {},
  context: {
    version: '',
    stateHandle: '',
    expiresAt: '',
    intent: '',
    currentAuthenticator: {
      type: '',
      value: {
        displayName: '',
        id: '',
        methods: [],
        key: '',
        type: '',
      },
    },
    authenticators: {
      type: '',
      value: [],
    },
    authenticatorEnrollments: {
      type: '',
      value: [],
    },
    enrollmentAuthenticator: {
      type: '',
      value: {
        displayName: '',
        id: '',
        methods: [],
        key: '',
        type: '',
      },
    },
    currentAuthenticatorEnrollment: {
      type: '',
      value: {
        displayName: '',
        id: '',
        methods: [],
        key: '',
        type: '',
      },
    },
    user: {
      type: '',
      value: {},
    },
    app: {
      type: '',
      value: {},
    },
  },
});

export const getStubTransactionWithNextStep = (stepName?: string): IdxTransaction => {
  const transaction = getStubTransaction();

  return { ...transaction, nextStep: { name: stepName ?? '' } };
};

export const getStubFormBag = (
  step?: string,
  orientation?: UISchemaLayoutType,
): FormBag => ({
  dataSchema: {
    submit: { step: step || IDX_STEP.CHALLENGE_AUTHENTICATOR },
    fieldsToTrim: [],
    fieldsToValidate: [],
    fieldsToExclude: () => ([]),
  } as DataSchemaBag,
  schema: {},
  uischema: {
    type: orientation || UISchemaLayoutType.VERTICAL,
    elements: [],
  },
  data: {},
});

export const getMockCreateCredentialsResponse = (): PublicKeyCredential => (
  {
    id: 'test',
    type: 'test',
    authenticatorAttachment: null,
    rawId: new ArrayBuffer(10),
    getClientExtensionResults: jest.fn(),
    response: {
      clientDataJSON: new ArrayBuffer(10),
      attestationObject: new ArrayBuffer(10),
    } as AuthenticatorAttestationResponse,
  } as PublicKeyCredential
);
