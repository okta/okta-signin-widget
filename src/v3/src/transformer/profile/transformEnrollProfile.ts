/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IDX_STEP, PASSWORD_REQUIREMENT_VALIDATION_DELAY_MS } from '../../constants';
import {
  ButtonElement,
  ButtonType,
  DescriptionElement,
  DividerElement,
  FieldElement,
  FormBag,
  IdxStepTransformer,
  LinkElement,
  PasswordRequirementsData,
  PasswordRequirementsElement,
  TitleElement,
  UISchemaElement,
  UISchemaLayoutType,
  WidgetMessage,
} from '../../types';
import {
  buildPasswordRequirementNotMetErrorList,
  getUserInfo,
  loc,
  updatePasswordRequirementsNotMetMessage,
  validatePassword,
} from '../../util';
import { buildPasswordRequirementListItems } from '../password';
import { getUIElementWithName } from '../utils';

export const transformEnrollProfile: IdxStepTransformer = ({ transaction, formBag }) => {
  const {
    availableSteps,
    nextStep: { name: stepName, inputs } = {},
    context,
    neededToProceed,
  } = transaction;
  // @ts-ignore OKTA-538692 uiDisplay missing from interface
  const { uiDisplay: { value: { label, buttonLabel } = {} } = {} } = context;
  const currentRemediation = neededToProceed.find((remediation) => remediation.name === stepName);
  const isUpdateProfile = currentRemediation?.href?.endsWith('idp/idx/enroll/update');
  const { dataSchema, uischema } = formBag;
  const userInfo = getUserInfo(transaction);
  const credentialsInput = inputs?.find((input) => input.name === 'credentials');
  const passwordSettings = (
    // @ts-ignore OKTA-545082 relatesTo prop missing from Input interface
    credentialsInput?.relatesTo?.value?.settings || {}
  ) as PasswordRequirementsData;
  const requirements = buildPasswordRequirementListItems(passwordSettings);

  // If passcode exists in elements, add password requirements element to Ui
  const passwordElement = getUIElementWithName(
    'credentials.passcode',
    uischema.elements as UISchemaElement[],
  ) as FieldElement;
  if (passwordElement) {
    passwordElement.options = {
      ...passwordElement.options,
      attributes: {
        ...passwordElement.options?.attributes,
        autocomplete: 'new-password',
      },
    };
    // @ts-ignore TODO: OKTA-539834 - messages missing from type
    const passwordErrors = passwordElement.options.inputMeta.messages?.value;
    if (passwordErrors?.length) {
      const messages = updatePasswordRequirementsNotMetMessage(passwordErrors);
      // @ts-ignore TODO: OKTA-539834 - messages missing from type
      passwordElement.options.inputMeta.messages.value = messages;
    }
    const passwordRequirementsElement: PasswordRequirementsElement = {
      type: 'PasswordRequirements',
      options: {
        id: 'password-authenticator--list',
        header: loc('password.complexity.requirements.header', 'login'),
        userInfo,
        settings: passwordSettings,
        requirements,
        validationDelayMs: PASSWORD_REQUIREMENT_VALIDATION_DELAY_MS,
      },
    };
    if (Object.keys(passwordSettings)?.length) {
      uischema.elements.unshift(passwordRequirementsElement);
    }

    // Controls form submission validation
    dataSchema['credentials.passcode'] = {
      validate: (data: FormBag['data']) => {
        const newPw = data['credentials.passcode'] as string;
        const errorMessages: WidgetMessage[] = [];
        if (!newPw && passwordElement.options.inputMeta.required) {
          errorMessages.push({
            class: 'ERROR',
            message: loc('model.validation.field.blank', 'login'),
            i18n: { key: 'model.validation.field.blank' },
          });
        }
        if (newPw) {
          const validations = validatePassword(newPw, userInfo, passwordSettings);
          const requirementNotMetMessages = buildPasswordRequirementNotMetErrorList(
            requirements,
            validations,
            'credentials.passcode',
          );
          errorMessages.push(...requirementNotMetMessages);
        }
        return errorMessages.length > 0 ? errorMessages : undefined;
      },
    };
  }

  const titleElement: TitleElement = {
    type: 'Title',
    options: { content: loc('oie.registration.form.title', 'login') },
  };

  const subtitleElement: DescriptionElement = {
    type: 'Description',
    contentType: 'subtitle',
    options: { content: loc('oie.form.field.optional.description', 'login') },
  };

  const submitBtnElement: ButtonElement = {
    type: 'Button',
    label: loc('oie.registration.form.submit', 'login'),
    options: {
      type: ButtonType.SUBMIT,
      step: transaction.nextStep!.name,
    },
  };

  if (isUpdateProfile) {
    if (label) {
      titleElement.options.content = loc('oie.registration.form.customize.label', 'login', [label]);
    } else {
      titleElement.options.content = loc('oie.primaryauth.submit', 'login');
    }

    if (buttonLabel) {
      submitBtnElement.label = loc('oie.registration.form.customize.buttonLabel', 'login', [buttonLabel]);
    } else {
      submitBtnElement.label = loc('oie.registration.form.update.submit', 'login');
    }
  }

  uischema.elements.unshift(subtitleElement);
  uischema.elements.unshift(titleElement);
  uischema.elements.push(submitBtnElement);

  const selectIdentifyStep = availableSteps?.find(({ name }) => name === IDX_STEP.SELECT_IDENTIFY);
  if (selectIdentifyStep) {
    uischema.elements.push({ type: 'Divider' } as DividerElement);
    const { name: step } = selectIdentifyStep;
    const signinLink: LinkElement = {
      type: 'Link',
      contentType: 'footer',
      options: {
        label: loc('signin', 'login'),
        step,
      },
    };
    const signinLabel: DescriptionElement = {
      type: 'Description',
      options: {
        content: loc('haveaccount', 'login'),
        dataSe: 'haveaccount',
      },
    };
    const loginEntryLayout = {
      type: UISchemaLayoutType.HORIZONTAL,
      elements: [signinLabel, signinLink],
    };

    uischema.elements.push(loginEntryLayout);
  }

  return formBag;
};
