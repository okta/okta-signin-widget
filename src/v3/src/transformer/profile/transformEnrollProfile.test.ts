/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxAuthenticator } from '@okta/okta-auth-js';
import { IDX_STEP, PASSWORD_REQUIREMENT_VALIDATION_DELAY_MS } from 'src/constants';
import { getStubFormBag, getStubTransactionWithNextStep } from 'src/mocks/utils/utils';
import {
  ButtonElement,
  ButtonType,
  DescriptionElement,
  FieldElement,
  LinkElement,
  PasswordRequirementsElement,
  TitleElement,
  UISchemaLayout,
  WidgetProps,
} from 'src/types';

import { transformEnrollProfile } from './transformEnrollProfile';

describe('Enroll Profile Transformer Tests', () => {
  const transaction = getStubTransactionWithNextStep();
  const formBag = getStubFormBag();
  let widgetProps: WidgetProps;
  beforeEach(() => {
    formBag.uischema.elements = [
      { type: 'Field', options: { inputMeta: { name: 'userProfile.firstName' } } } as FieldElement,
      { type: 'Field', options: { inputMeta: { name: 'userProfile.lastName' } } } as FieldElement,
      { type: 'Field', options: { inputMeta: { name: 'userProfile.email' } } } as FieldElement,
    ];
    transaction.nextStep = {
      name: '',
    };
    widgetProps = {};
  });

  it('should only add title and submit button when select-identify doesnt exist in available steps '
    + 'and passcode element doesnt exist in schema', () => {
    const updatedFormBag = transformEnrollProfile({ transaction, formBag, widgetProps });

    expect(updatedFormBag).toMatchSnapshot();
    expect(updatedFormBag.uischema.elements.length).toBe(6);
    expect(updatedFormBag.uischema.elements[0].type).toBe('Title');
    expect((updatedFormBag.uischema.elements[0] as TitleElement).options?.content)
      .toBe('oie.registration.form.title');
    expect((updatedFormBag.uischema.elements[1] as DescriptionElement).options?.content)
      .toBe('oie.form.field.optional.description');
    expect((updatedFormBag.uischema.elements[2] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.firstName');
    expect((updatedFormBag.uischema.elements[3] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.lastName');
    expect((updatedFormBag.uischema.elements[4] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.email');
    expect((updatedFormBag.uischema.elements[5] as ButtonElement).label)
      .toBe('oie.registration.form.submit');
    expect((updatedFormBag.uischema.elements[5] as ButtonElement).type).toBe('Button');
    expect((updatedFormBag.uischema.elements[5] as ButtonElement).options?.type)
      .toBe(ButtonType.SUBMIT);
  });

  it('should add password requirements along with title, and submit button when passcode exists but password settings are empty', () => {
    formBag.uischema.elements.push({
      type: 'Field',
      label: 'Password',
      options: {
        inputMeta: { name: 'credentials.passcode', secret: true },
      },
    } as FieldElement);
    const mockUserInfo = {
      identifier: 'testuser@okta.com',
      profile: { firstName: 'test', lastName: 'user' },
    };
    transaction.context.user = {
      type: 'object',
      value: mockUserInfo,
    };

    const updatedFormBag = transformEnrollProfile({ transaction, formBag, widgetProps });

    expect(updatedFormBag).toMatchSnapshot();
    expect(updatedFormBag.uischema.elements.length).toBe(7);
    expect(updatedFormBag.uischema.elements[0].type).toBe('Title');
    expect((updatedFormBag.uischema.elements[0] as TitleElement).options?.content)
      .toBe('oie.registration.form.title');
    expect((updatedFormBag.uischema.elements[1] as DescriptionElement).options?.content)
      .toBe('oie.form.field.optional.description');
    expect((updatedFormBag.uischema.elements[2] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.firstName');
    expect((updatedFormBag.uischema.elements[3] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.lastName');
    expect((updatedFormBag.uischema.elements[4] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.email');
    expect((updatedFormBag.uischema.elements[5] as FieldElement).options?.inputMeta.name)
      .toBe('credentials.passcode');
    expect((updatedFormBag.uischema.elements[5] as FieldElement).options?.attributes?.autocomplete)
      .toBe('new-password');
    expect((updatedFormBag.uischema.elements[6] as ButtonElement).label)
      .toBe('oie.registration.form.submit');
    expect((updatedFormBag.uischema.elements[6] as ButtonElement).type).toBe('Button');
    expect((updatedFormBag.uischema.elements[6] as ButtonElement).options?.type)
      .toBe(ButtonType.SUBMIT);
  });

  it('should add password requirements along with title, and submit button when passcode and password settings exists', () => {
    formBag.uischema.elements.push({
      type: 'Field',
      label: 'Password',
      options: {
        inputMeta: { name: 'credentials.passcode', secret: true },
      },
    } as FieldElement);
    transaction.nextStep = {
      name: '',
      inputs: [
        { name: 'userProfile', value: [] },
        {
          name: 'credentials',
          value: [],
          // @ts-ignore OKTA-545082 relatesTo prop missing from Input interface
          relatesTo: {
            value: {
              settings: { complexity: { minNumber: 1, minSymbol: 1 } },
            } as unknown as IdxAuthenticator,
          },
        },
      ],
    };
    const mockUserInfo = {
      identifier: 'testuser@okta.com',
      profile: { firstName: 'test', lastName: 'user' },
    };
    transaction.context.user = {
      type: 'object',
      value: mockUserInfo,
    };

    const updatedFormBag = transformEnrollProfile({ transaction, formBag, widgetProps });

    expect(updatedFormBag).toMatchSnapshot();
    expect(updatedFormBag.uischema.elements.length).toBe(8);
    expect(updatedFormBag.uischema.elements[0].type).toBe('Title');
    expect((updatedFormBag.uischema.elements[0] as TitleElement).options?.content)
      .toBe('oie.registration.form.title');
    expect((updatedFormBag.uischema.elements[1] as DescriptionElement).options?.content)
      .toBe('oie.form.field.optional.description');
    expect(updatedFormBag.uischema.elements[2].type).toBe('PasswordRequirements');
    expect((updatedFormBag.uischema.elements[2] as PasswordRequirementsElement).options?.id)
      .toBe('password-authenticator--list');
    expect((updatedFormBag.uischema.elements[2] as PasswordRequirementsElement).options?.header)
      .toBe('password.complexity.requirements.header');
    expect((updatedFormBag.uischema.elements[2] as PasswordRequirementsElement).options?.userInfo)
      .toEqual(mockUserInfo);
    expect((updatedFormBag.uischema.elements[2] as PasswordRequirementsElement).options?.settings)
      .toEqual({ complexity: { minNumber: 1, minSymbol: 1 } });
    expect((updatedFormBag.uischema.elements[2] as PasswordRequirementsElement)
      .options?.validationDelayMs).toBe(PASSWORD_REQUIREMENT_VALIDATION_DELAY_MS);
    expect((updatedFormBag.uischema.elements[3] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.firstName');
    expect((updatedFormBag.uischema.elements[4] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.lastName');
    expect((updatedFormBag.uischema.elements[5] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.email');
    expect((updatedFormBag.uischema.elements[6] as FieldElement).options?.inputMeta.name)
      .toBe('credentials.passcode');
    expect((updatedFormBag.uischema.elements[6] as FieldElement).options?.attributes?.autocomplete)
      .toBe('new-password');
    expect((updatedFormBag.uischema.elements[7] as ButtonElement).label)
      .toBe('oie.registration.form.submit');
    expect((updatedFormBag.uischema.elements[7] as ButtonElement).type).toBe('Button');
    expect((updatedFormBag.uischema.elements[7] as ButtonElement).options?.type)
      .toBe(ButtonType.SUBMIT);
  });

  it('should add link to log in when select-identify step exists in remediation', () => {
    transaction.availableSteps = [{
      name: IDX_STEP.SELECT_IDENTIFY,
      action: jest.fn(),
    }];

    const updatedFormBag = transformEnrollProfile({ transaction, formBag, widgetProps });

    expect(updatedFormBag).toMatchSnapshot();
    expect(updatedFormBag.uischema.elements.length).toBe(8);
    expect((updatedFormBag.uischema.elements[0] as TitleElement).options?.content)
      .toBe('oie.registration.form.title');
    expect((updatedFormBag.uischema.elements[1] as DescriptionElement).options?.content)
      .toBe('oie.form.field.optional.description');
    expect((updatedFormBag.uischema.elements[2] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.firstName');
    expect((updatedFormBag.uischema.elements[3] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.lastName');
    expect((updatedFormBag.uischema.elements[4] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.email');
    expect((updatedFormBag.uischema.elements[5] as ButtonElement).label)
      .toBe('oie.registration.form.submit');
    expect((updatedFormBag.uischema.elements[5] as ButtonElement).type).toBe('Button');
    expect((updatedFormBag.uischema.elements[5] as ButtonElement).options?.type)
      .toBe(ButtonType.SUBMIT);
    expect(updatedFormBag.uischema.elements[6].type).toBe('Divider');
    expect(updatedFormBag.uischema.elements[7].type).toBe('HorizontalLayout');
    expect(((updatedFormBag.uischema.elements[7] as UISchemaLayout)
      .elements[0] as DescriptionElement).options.content).toBe('haveaccount');
    expect(((updatedFormBag.uischema.elements[7] as UISchemaLayout)
      .elements[1] as LinkElement).options.label).toBe('signin');
  });
});
