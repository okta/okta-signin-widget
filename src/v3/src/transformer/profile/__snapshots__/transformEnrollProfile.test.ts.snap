// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enroll Profile Transformer Tests should add link to log in when select-identify step exists in remediation 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {
    "credentials.passcode": Object {
      "validate": [Function],
    },
    "fieldsToExclude": [Function],
    "fieldsToTrim": Array [],
    "fieldsToValidate": Array [],
    "submit": Object {
      "step": "challenge-authenticator",
    },
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "content": "oie.registration.form.title",
        },
        "type": "Title",
      },
      Object {
        "contentType": "subtitle",
        "options": Object {
          "content": "oie.form.field.optional.description",
        },
        "type": "Description",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.firstName",
          },
        },
        "type": "Field",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.lastName",
          },
        },
        "type": "Field",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.email",
          },
        },
        "type": "Field",
      },
      Object {
        "label": "oie.registration.form.submit",
        "options": Object {
          "step": "",
          "type": "submit",
        },
        "type": "Button",
      },
      Object {
        "type": "Divider",
      },
      Object {
        "elements": Array [
          Object {
            "options": Object {
              "content": "haveaccount",
              "dataSe": "haveaccount",
            },
            "type": "Description",
          },
          Object {
            "contentType": "footer",
            "options": Object {
              "label": "signin",
              "step": "select-identify",
            },
            "type": "Link",
          },
        ],
        "type": "HorizontalLayout",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`Enroll Profile Transformer Tests should add password requirements along with title, and submit button when passcode and password settings exists 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {
    "credentials.passcode": Object {
      "validate": [Function],
    },
    "fieldsToExclude": [Function],
    "fieldsToTrim": Array [],
    "fieldsToValidate": Array [],
    "submit": Object {
      "step": "challenge-authenticator",
    },
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "content": "oie.registration.form.title",
        },
        "type": "Title",
      },
      Object {
        "contentType": "subtitle",
        "options": Object {
          "content": "oie.form.field.optional.description",
        },
        "type": "Description",
      },
      Object {
        "options": Object {
          "header": "password.complexity.requirements.header",
          "id": "password-authenticator--list",
          "requirements": Array [
            Object {
              "label": "password.complexity.number.description",
              "ruleKey": "minNumber",
            },
            Object {
              "label": "password.complexity.symbol.description",
              "ruleKey": "minSymbol",
            },
          ],
          "settings": Object {
            "complexity": Object {
              "minNumber": 1,
              "minSymbol": 1,
            },
          },
          "userInfo": Object {
            "identifier": "testuser@okta.com",
            "profile": Object {
              "firstName": "test",
              "lastName": "user",
            },
          },
          "validationDelayMs": 50,
        },
        "type": "PasswordRequirements",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.firstName",
          },
        },
        "type": "Field",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.lastName",
          },
        },
        "type": "Field",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.email",
          },
        },
        "type": "Field",
      },
      Object {
        "label": "Password",
        "options": Object {
          "attributes": Object {
            "autocomplete": "new-password",
          },
          "inputMeta": Object {
            "name": "credentials.passcode",
            "secret": true,
          },
        },
        "type": "Field",
      },
      Object {
        "label": "oie.registration.form.submit",
        "options": Object {
          "step": "",
          "type": "submit",
        },
        "type": "Button",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`Enroll Profile Transformer Tests should not add password requirements but should add title, password, and button elements when passcode exists but password settings are empty 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {
    "credentials.passcode": Object {
      "validate": [Function],
    },
    "fieldsToExclude": [Function],
    "fieldsToTrim": Array [],
    "fieldsToValidate": Array [],
    "submit": Object {
      "step": "challenge-authenticator",
    },
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "content": "oie.registration.form.title",
        },
        "type": "Title",
      },
      Object {
        "contentType": "subtitle",
        "options": Object {
          "content": "oie.form.field.optional.description",
        },
        "type": "Description",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.firstName",
          },
        },
        "type": "Field",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.lastName",
          },
        },
        "type": "Field",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.email",
          },
        },
        "type": "Field",
      },
      Object {
        "label": "Password",
        "options": Object {
          "attributes": Object {
            "autocomplete": "new-password",
          },
          "inputMeta": Object {
            "name": "credentials.passcode",
            "secret": true,
          },
        },
        "type": "Field",
      },
      Object {
        "label": "oie.registration.form.submit",
        "options": Object {
          "step": "",
          "type": "submit",
        },
        "type": "Button",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`Enroll Profile Transformer Tests should only add title and submit button when select-identify doesnt exist in available steps and passcode element doesnt exist in schema 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {
    "fieldsToExclude": [Function],
    "fieldsToTrim": Array [],
    "fieldsToValidate": Array [],
    "submit": Object {
      "step": "challenge-authenticator",
    },
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "content": "oie.registration.form.title",
        },
        "type": "Title",
      },
      Object {
        "contentType": "subtitle",
        "options": Object {
          "content": "oie.form.field.optional.description",
        },
        "type": "Description",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.firstName",
          },
        },
        "type": "Field",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.lastName",
          },
        },
        "type": "Field",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.email",
          },
        },
        "type": "Field",
      },
      Object {
        "label": "oie.registration.form.submit",
        "options": Object {
          "step": "",
          "type": "submit",
        },
        "type": "Button",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;
